src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java

```diff
@@ -229,14 +227,31 @@ protected Object invokeStep(StepDescriptor d, String name, Object args) {
 
         if (ps.body == null && !hack) {
             an = new StepAtomNode(exec, d, thread.head.get());
-            // TODO: use CPS call stack to obtain the current call site source location. See JENKINS-23013
-            thread.head.setNewHead(an);
         } else {
             an = new StepStartNode(exec, d, thread.head.get());
-            thread.head.setNewHead(an);
         }
 
-        final CpsStepContext context = new CpsStepContext(d,thread,handle,an,ps.body);
+        // Ensure ArgumentsAction is attached before we notify even synchronous listeners:
+        final CpsStepContext context = new CpsStepContext(d, thread, handle, an, ps.body);
+        try {
```

Comment:
I *think* that you want the classloader switch to happen before the try block for safety, no?  Or is there a reason why we shouldn't continue running this logic under the `CpsVmExecutorService.ORIGINAL_CONTEXT_CLASS_LOADER.get()` anymore?

Reply:
It never needed to be run with a special CCL to begin with. That is just where you happened to drop that new code. The CCL is for user code, mainly `StepExecution.start` implementations.

Reply:
The `ORIGINAL_CONTEXT_CLASSLOADER` is for non-Groovy code though IIRC -- and IIRC we are switching away from the GroovyClassLoader to avoid stuff being loaded there (and potential memory leaks or issues with classloaders not matching).

I *think* you still want this to be happening with the non-Groovy classloader.

Reply:
> The `ORIGINAL_CONTEXT_CLASSLOADER` is for non-Groovy code though

For non-Groovy _plugin_ code, yes.

> we are switching away from the GroovyClassLoader to avoid stuff being loaded there (and potential memory leaks or issues with classloaders not matching)

We are switching to the CCL used for general plugin Java code while we run the general plugin Java code: `StepDescriptor.newInstance`, `Step.start`, and (especially) `StepExecution.start`. `ArgumentsActionImpl` has nothing to do with this.

Reply:
FTR this system is from #50. #98 happened to drop `ArgumentsActionImpl` into the same `try` block as a convenience AFAICT—because it was checking for `instanceof ParallelStep`, which can only be done after instantiating the `Step`. This PR instead checks the `StepDescriptor`, which can be done earlier.